Source/WebCore/style/StyleInvalidator.cpp

```diff
@@ -165,11 +172,15 @@ Invalidator::CheckDescendants Invalidator::invalidateIfNeeded(Element& element,
     case Validity::Valid:
     case Validity::AnimationInvalid:
     case Validity::InlineStyleInvalid: {
-        for (auto& ruleSet : m_ruleSets) {
-            ElementRuleCollector ruleCollector(element, *ruleSet, selectorMatchingState);
+        for (auto& ruleSetWithNegation : m_ruleSets) {
+            ElementRuleCollector ruleCollector(element, *ruleSetWithNegation.ruleSet, selectorMatchingState);
             ruleCollector.setMode(SelectorChecker::Mode::CollectingRulesIgnoringVirtualPseudoElements);
-
-            if (ruleCollector.matchesAnyAuthorRules()) {
+            auto matchesAnyRules = ruleCollector.matchesAnyAuthorRules();
+            if (matchesAnyRules) {
+                element.invalidateStyleInternal();
+                break;
+            }
+            if (!matchesAnyRules && ruleSetWithNegation.isNegation == IsNegation::Yes) {
                 element.invalidateStyleInternal();
```

Comment:
This is not correct and just papers over the problem by invalidating unnecessarily. It will cause massive over-invalidation for :not selectors.

We try invalidation both before and after the state change (ChildChangeInvalidation::invalidateForHasBefore/AfterMutation) and that code should take care of negations.